Update package docs to include package bundle information (#6247)
aws/eks-anywhere: docs/content/en/docs/packages/packagecontroller.md

---

> title: "Manging the package controller"

Reviewer: ```suggestion
title: "Managing the package controller"
```

linkTitle: "Package controller"
weight: 4
aliases:
    /docs/tasks/packages/packagecontroller/
---

### Installing the package controller

{{% alert title="Important" color="warning" %}}

> The package controller installation creates a package bundle controller resource for each cluster, thus allowing each to activate different package bundle version. Ideally, you should never delete this resource because it would mean losing that information and upon re-installing, the latest bundle will be selected. However, you can always go back to the previous bundle version. For more information, see [Managing package bundles.]({{< relref "./packagebundles" >}})

Reviewer 2: ```suggestion
The package controller installation creates a package bundle controller resource for each cluster, thus allowing each to activate a different package bundle version. Ideally, you should never delete this resource because it would mean losing that information and upon re-installing, the latest bundle would be selected. However, you can always go back to the previous bundle version. For more information, see [Managing package bundles.]({{< relref "./packagebundles" >}})
```

{{% /alert %}}

The package controller is typically installed during cluster creation, but may be disabled intentionally in your `cluster.yaml` by setting `spec.packages.disable` to `true`.

If you created a cluster without the package controller or if the package controller was not properly configured, you may need to manually install it.

1. Enable the package controller in your `cluster.yaml`, if it was previously disabled:
    ```yaml
    apiVersion: anywhere.eks.amazonaws.com/v1alpha1
    kind: Cluster
    metadata:
      name: mgmt
    spec:
      packages:
        disable: false
    ```

2. Make sure you are authenticated with the AWS CLI. Use the credentials you set up for packages. These credentials should have [limited capabilities]({{< relref "./prereq#setup-authentication-to-use-curated-packages" >}}):

    ```bash
    export AWS_ACCESS_KEY_ID="your*access*id"
    export AWS_SECRET_ACCESS_KEY="your*secret*key"
    export EKSA_AWS_ACCESS_KEY_ID="your*access*id"
    export EKSA_AWS_SECRET_ACCESS_KEY="your*secret*key"
    ```

3. Verify your credentials are working:
    ```shell
    aws sts get-caller-identity
    ```

> 4. Authenticate docker to the private AWS ECR registry on account `783794618700` with your AWS credentials. It houses the EKS Anywhere packages artifacts and authentication is required to pull imagaes from it.

Reviewer 2: ```suggestion
4. Authenticate docker to the private AWS ECR registry on account `783794618700` with your AWS credentials. It houses the EKS Anywhere packages artifacts. Authentication is required to pull images from it.
```

    ```bash
    aws ecr get-login-password | docker login --username AWS --password-stdin 783794618700.dkr.ecr.us-west-2.amazonaws.com
    ```

5. Verify you can pull an image from the packages registry:
    ```bash
    docker pull 783794618700.dkr.ecr.us-west-2.amazonaws.com/emissary-ingress/emissary:v3.5.1-bf70150bcdfe3a5383ec8ad9cd7eea801a0cb074
    ```
    If the image downloads successfully, it worked!

> 6. Now, install the package controller using the eks anywhere package CLI:

Reviewer 2: ```suggestion
6. Now, install the package controller using the eks anywhere package CLI:
```
Is the package literally named "eks anywhere". If not change that to "EKS Anywhere".

Author: Yes, EKS Anywhere makes sense here. It’s referring to the EKS Anywhere Packages CLI

    ```shell
    eksctl anywhere install packagecontroller -f cluster.yaml
    ```

    The package controller should now be installed!

> 7. Verify, use kubectl to check the eks-anywhere-packages pod is running in your management cluster:

Reviewer 2: ```suggestion
7. Use kubectl to check the eks-anywhere-packages pod is running in your management cluster:
```

    ```
    kubectl get pods -n eksa-packages 
    NAME                                     READY   STATUS    RESTARTS   AGE
    eks-anywhere-packages-55bc54467c-jfhgp   1/1     Running   0          21s
    ```

### Updating the package credentials

You may need to create or update your credentials which you can do with a command like this. Set the environment variables to the proper values before running the command.
  ```bash
  kubectl delete secret -n eksa-packages aws-secret
  kubectl create secret -n eksa-packages generic aws-secret \
    --from-literal=AWS_ACCESS_KEY_ID=${EKSA_AWS_ACCESS_KEY_ID} \
    --from-literal=AWS_SECRET_ACCESS_KEY=${EKSA_AWS_SECRET_ACCESS_KEY}  \
    --from-literal=REGION=${EKSA_AWS_REGION}
  ```

### Upgrade the packages controller

EKS Anywhere v0.15.0 (packages controller v0.3.9+) and onwards includes support for eks-anywhere-packages controller as a self-managed package feature. The package controller now  upgrades automatically according to the version specified within the management cluster's selected package bundle.

For any version prior to v0.3.X, manual steps must be executed to upgrade.

{{% alert title="Important" color="warning" %}}

> This operation may change your cluster's selected package bundle to the latest, however, you can always go back to the previous bundle version. For more information, see [Managing package bundles.]({{< relref "./packagebundles" >}})

Reviewer 2: ```suggestion
This operation may change your cluster's selected package bundle to the latest version. However, you can always go back to the previous bundle version. For more information, see [Managing package bundles.]({{< relref "./packagebundles" >}})
```

---
title: "Manging the package controller"
linkTitle: "Package controller"
weight: 4
aliases:
    /docs/tasks/packages/packagecontroller/
---

> ### Install package controller after installation

Reviewer 3: this is kinda confusing. Do you mean "Reinstall package controller"?

Author: I moved this from the prequisite page, so that is correct. However, it is the same process for both. I can probably clarify this.

I think I can just name this section "Installing the package controller" and then add context below as to when this would be used.

{{% alert title="Important" color="warning" %}}
**Recommendation**: Ideally, you should never delete package controller after installation. When the package controller is re-installed, it will try to pull and activate the latest bundle. So, there is a risk that your packages will be upgraded to a different version. See more information about package bundles [here]({{< relref "./packagebundles" >}})

> {{% /alert %}}

Reviewer 4: Installed packages won't be automatically upgraded even if a different bundle gets activated in the pbc resource. Actually, the risk is if the pbc resource is deleted and then gets recreated, the active bundle recorded in the pbc resource will point to the latest bundle.

Reviewer 4: but you can always get back to the previous bundle version by running `eksctl anywhere upgrade packages --bundle-version xxx --cluster xxx`.

> **Recommendation**: Ideally, you should never delete package controller after installation. When the package bundle controller resource is deleted and then gets recreated, the active bundle recorded in the package bundle controller resource will point to the latest bundle, but you can always go back to the previous bundle version. See more information about managing package bundles [here]({{< relref "./packagebundles" >}})

Reviewer 3: it's we are not supposed to delete package controller, or package bundle controller CR? Which one will trigger the active bundle to point to latest?

Author: I'll move this down to the section about manually upgrading the package controller, where this is more relevant

If you created a cluster without the package controller or if the package controller was not properly configured, you may need to do some things to enable it.

Make sure you are authenticated with the AWS CLI. Use the credentials you set up for packages. These credentials should have [limited capabilities]({{< relref "./prereq#setup-authentication-to-use-curated-packages" >}}):

```bash
export AWS_ACCESS_KEY_ID="your*access*id"
export AWS_SECRET_ACCESS_KEY="your*secret*key"
export EKSA_AWS_ACCESS_KEY_ID="your*access*id"
export EKSA_AWS_SECRET_ACCESS_KEY="your*secret*key"
```

Verify your credentials are working:
```shell
aws sts get-caller-identity
```
Login to docker
```bash

> aws ecr get-login-password |docker login --username AWS --password-stdin 783794618700.dkr.ecr.us-west-2.amazonaws.com

Reviewer 3: what is 783794618700 account?

Reviewer 2: It looks like you are using aws credentials to have docker login to ECR to gain access to the registry that holds the EKS Anywhere packages artifacts. If that is true, could you say that on the "Login to docker" line? Also, Can you make these steps into a numbered list?

Author: @jiayiwang7 It houses the EKS Anywhere packages artifacts. I'll update the text here to give that information

```

Verify you can pull an image
```bash
docker pull 783794618700.dkr.ecr.us-west-2.amazonaws.com/emissary-ingress/emissary:v3.5.1-bf70150bcdfe3a5383ec8ad9cd7eea801a0cb074
```
If the image downloads successfully, it worked!

If you do not have the package controller installed (it is installed by default), install it now:
```shell
eksctl anywhere install packagecontroller -f cluster.yaml
```

> If you had the package controller disabled, you may need to modify your `cluster.yaml` to enable it.

Reviewer 3: and run eks-a cli upgrade?

Author: No, this is suppose to be a follow up if you had it disabled in your `cluster.yaml` it seems. I can move this further up in the steps to make this more clear.

```yaml
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: Cluster
metadata:
  name: billy
spec:
  packages:
    disable: false
```

You may need to create or update your credentials which you can do with a command like this. Set the environment variables to the proper values before running the command.
```bash
kubectl delete secret -n eksa-packages aws-secret
kubectl create secret -n eksa-packages generic aws-secret \
   --from-literal=AWS_ACCESS_KEY_ID=${EKSA_AWS_ACCESS_KEY_ID} \
   --from-literal=AWS_SECRET_ACCESS_KEY=${EKSA_AWS_SECRET_ACCESS_KEY}  \
   --from-literal=REGION=${EKSA_AWS_REGION}
```

### Upgrade the packages controller

{{% alert title="Important" color="warning" %}}

> EKS Anywhere v0.15.0 (packages controller v0.3.9+) and onwards includes support for eks-anywhere-packages controller as a self-managed package feature. The package controller now  upgrades automatically according to the version within the selected package bundle. For any version prior to v0.3.X, manual steps must be executed to upgrade.

Reviewer 2: ```suggestion
EKS Anywhere v0.15.0 (packages controller v0.3.9+) and onwards includes support for eks-anywhere-packages controller as a self-managed package feature. The package controller now upgrades automatically according to the version within the selected package bundle. For any version prior to v0.3.X, manual steps must be executed to upgrade.
```

{{% /alert %}}

> To manually upgrade the package controller,

Reviewer 3: does this operation change the active bundle to latest and update existing packages (potentially) to latest? If so can we add warning?

Author: It may change the active bundle, but does not update the existing packages. See [here](https://github.com/aws/eks-anywhere/pull/6247#discussion_r1271122150). I will add the warning about the changing active bundle.

> To manually upgrade the package controller,

Reviewer 2: ```suggestion
To manually upgrade the package controller, do the following:
```

1. Ensure the namespace will be kept
```
kubectl annotate namespaces eksa-packages helm.sh/resource-policy=keep
```

2. Uninstall the eks-anywhere-package helm release
```
helm uninstall eks-anywhere-packages
```

3. Remove the secret called aws-secret (we will need credentials when installing the new version)
```
kubectl delete secret -n eksa-package aws-secret
```

4. Install the new version using the latest eksctl-anywhere binary
```

> eksctl anywhere install packagecontroller -f ${CLUSTER_NAME}.yaml

Reviewer 2: Is this the management or workload CLUSTER_NAME?

Author: The management cluster. I will clarify that
